Generate Alias for RLWE, ModArith, and Polynomial

ZenithalHourlyRate opened the pull request:
Check https://github.com/google/heir/pull/1130#issuecomment-2508866775

Cc @j2kun Definitely worth a blog post. Never seen a tutorial on this.

With this change, `--mlir-to-openfhe-bgv` result would be much cleaner.

<del>Though a lot of tests failed due to this change...</del>Can be mitigated by `--mlir-print-local-scope`

```mlir
!Z463187969_i32_ = !mod_arith.int<463187969 : i32>
#ring_Z463187969_i32_1_x8_ = #polynomial.ring<coefficientType = !Z463187969_i32_, polynomialModulus = <1 + x**8>>
!rlwe_pt = !lwe.rlwe_plaintext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>, ring = #ring_Z463187969_i32_1_x8_, underlying_type = i16>
!rlwe_ct = !lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>, rlwe_params = <ring = #ring_Z463187969_i32_1_x8_>, underlying_type = i16>
!rlwe_ct1 = !lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>, rlwe_params = <dimension = 3, ring = #ring_Z463187969_i32_1_x8_>, underlying_type = i16>
module {
  func.func @func(%arg0: !openfhe.crypto_context, %arg1: !rlwe_ct, %arg2: !rlwe_ct) -> !rlwe_ct {
    %0 = openfhe.mul_no_relin %arg0, %arg1, %arg1 : (!openfhe.crypto_context, !rlwe_ct, !rlwe_ct) -> !rlwe_ct1
    %1 = openfhe.relin %arg0, %0 : (!openfhe.crypto_context, !rlwe_ct1) -> !rlwe_ct
    return %1 : !rlwe_ct
  }
  func.func @func__encrypt__arg0(%arg0: !openfhe.crypto_context, %arg1: i16, %arg2: !openfhe.public_key) -> !rlwe_ct {
    %splat = tensor.splat %arg1 : tensor<8xi16>
    %0 = arith.extsi %splat : tensor<8xi16> to tensor<8xi64>
    %1 = openfhe.make_packed_plaintext %arg0, %0 : (!openfhe.crypto_context, tensor<8xi64>) -> !rlwe_pt
    %2 = openfhe.encrypt %arg0, %1, %arg2 : (!openfhe.crypto_context, !rlwe_pt, !openfhe.public_key) -> !rlwe_ct
    return %2 : !rlwe_ct
  }
```

j2kun reviewed (approved):
LGTM! Care to add a test that the aliases are printed correctly? It is aesthetic truly, but a small one might be good to ensure someone doesn't accidentally remove this feature.

j2kun commented on lib/Dialect/LWE/IR/LWEDialect.cpp:
```diff
@@ -34,6 +34,25 @@ namespace mlir {
+                   })
+                   .Case<RLWEPlaintextType>([&](auto& type) {
+                     os << "rlwe_pt";
+                     return AliasResult::OverridableAlias;
```
Is there a reason these are overridable while the others are final?

ZenithalHourlyRate replied:
>  enum class AliasResult {
    /// The object (type or attribute) is not supported by the hook
    /// and an alias was not provided.
    NoAlias,
    /// An alias was provided, but it might be overriden by other hook.
    OverridableAlias,
    /// An alias was provided and it should be used
    /// (no other hooks will be checked).
    FinalAlias
  };
  
I thought that if two different RLWECiphertextType has the same alias with `FinalAlias`, MLIR would complain, as by `(no other hooks will be checked)` it wont assign number to them. To reply this comment I expreimented with using `FinalAlias`, found that MLIR still assign number to them. I think we can all use `FinalAlias` safely.

ZenithalHourlyRate commented:
> Care to add a test that the aliases are printed correctly? It is aesthetic truly, but a small one might be good to ensure someone doesn't accidentally remove this feature.

Added three alias.mlir test. Fixed the lower_mul test as it can not use `mlir-print-local-scope` as `affine_map` is already an alias. lower_ntt_perf_runner should be fixed in another PR.

ZenithalHourlyRate commented:
Rebased to pass the CI.

AlexanderViand-Intel reviewed (approved):
This is super cool, thanks! I also agree, this is definitively worth documenting @j2kun 👍 

@ZenithalHourlyRate Do you know what's up with the trailing underscores in the alias names?

ZenithalHourlyRate commented:
> @ZenithalHourlyRate Do you know what's up with the trailing underscores in the alias names?

MLIR wants to ensure the name does not conflict with auto-generated one ending with digits so added a underscore to _sanitize_ it

https://github.com/llvm/llvm-project/blob/637a1ae85538554b98973861802e842fc30d0f91/mlir/lib/IR/AsmPrinter.cpp#L1030-L1036

asraa reviewed (approved):
Thanks! This is really cool. Thanks for figuring this out - hopefully we can add these to the new types too later.

AlexanderViand-Intel commented:
I might have found a bug/unexpected alias while playing around with RNS stuff:

```llvm
!a = !mod_arith.int<463187969 : i32>
!b = !mod_arith.int<463187970 : i32>
!c = !mod_arith.int<463187971 : i32>
!rns = !rns.rns<!a, !b, !c>
#ring = #polynomial.ring<coefficientType = !rns, polynomialModulus = <1 + x**8>>
!rlwe_ct = !lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>, rlwe_params = <ring = #ring>, underlying_type = tensor<8xi16>>
module {
  func.func @dot_product(%arg0: !rlwe_ct, %arg1: !rlwe_ct) -> !rlwe_ct {  
    %0 = bgv.add %arg0, %arg1 : !rlwe_ct
    return %0 : !rlwe_ct
  }
}
``` 

roundtrips to
```llvm
!Z463187970_i32_ = !mod_arith.int<463187970 : i32>
!Z463187971_i32_ = !mod_arith.int<463187971 : i32>
!ring_Z463187969_i32_ = !mod_arith.int<463187969 : i32>
!rlwe_ct = !lwe.rlwe_ciphertext<encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>, rlwe_params = <ring = <coefficientType = !rns.rns<!ring_Z463187969_i32_, !Z463187970_i32_, !Z463187971_i32_>, polynomialModulus = <1 + x**8>>>, underlying_type = tensor<8xi16>>
module {
  func.func @dot_product(%arg0: !openfhe.crypto_context, %arg1: !rlwe_ct, %arg2: !rlwe_ct) -> !rlwe_ct {
    %0 = openfhe.add %arg0, %arg1, %arg2 : (!openfhe.crypto_context, !rlwe_ct, !rlwe_ct) -> !rlwe_ct
    return %0 : !rlwe_ct
  }
}
```

Notice how the first `mod_arith.int` in the RNS somehow gets the alias `!ring_...` rather than `!Z.....`. 

Obviously not a showstopper, but I figured I'd ask if either of you @ZenithalHourlyRate  or @j2kun knows a quick fix for this, rather than trying to figure this out myself :)

ZenithalHourlyRate commented:
I'm looking into it.
